File: 21-merge-two-sorted-lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:

        def outBounds(temp, curr):
            while temp:
                curr.next = temp
                temp = temp.next
                curr = curr.next

        temp1, temp2 = list1, list2
        dummy = ListNode()
        curr = dummy

        while temp1 and temp2:
            if temp1.val <= temp2.val:
                curr.next = temp1
                temp1 = temp1.next
            else:
                curr.next = temp2
                temp2 = temp2.next          
            curr = curr.next
        if temp1:
            outBounds(temp1,curr) 
        else:
            outBounds(temp2, curr)   

        return dummy.next             

File: 21-merge-two-sorted-lists/merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {

    dummy := &ListNode{-1, nil}
    curr := dummy
    for list1 != nil || list2 != nil {
        if list1 != nil && list2 != nil {
            if list1.Val <= list2.Val {
                curr.Next = list1
                list1 = list1.Next
            } else {
                curr.Next = list2
                list2 = list2.Next
            }
        } else if list1 != nil {
            curr.Next = list1
            list1 = list1.Next
        } else {
            curr.Next = list2
            list2 = list2.Next
        }

        curr = curr.Next
    }

    return dummy.Next
    
}